syntax = "proto3";
option optimize_for = CODE_SIZE;

package RPC;

message VariantParam
{
    oneof Value
    {
        fixed32 Fixed32=1;
        sfixed32 SFixed32=2;
        fixed64 Fixed64=3;
        sfixed64 SFixed64=4;
        bool    Bool=5;
        float   Float=6;
        double  Double=7;
        string  String=8;
        string  Decimal=9;
        bytes   Bytes=10;
        bytes   Guid=11;
    }
};

message RpcInvoke
{
    sfixed32 ObjectID=1;
    fixed32 InterfaceID=2;
    sfixed32 CallID=3;
    int32 DispID=4;
    int32 wFlags=5;
    repeated VariantParam Params=6;
}

message RpcReturn
{
    sfixed32 CallID=1;
    sfixed32 HResult=2;
    repeated VariantParam Ref_Out_Params=3;

    message ExceptionInfo
    {
        string Description=1;
        int32 scode=2;
        int32 wCode=3;
    }

    ExceptionInfo Exception=4;
}
